worker/src/ee/experiments/experimentService.ts

```diff
@@ -0,0 +1,259 @@
+  variables: string[],
+): { role: string; content: string }[] => {
+  const processContent = (content: string) =>
+    content.replace(/\{\{(\w+)\}\}/g, (_, variable) =>
+      variables.includes(variable)
+        ? (itemInput[variable] ?? "")
+        : `{{${variable}}}`,
+    );
+
+  if (typeof prompt === "string") {
+    return [{ role: ChatMessageRole.System, content: processContent(prompt) }];
+  } else {
+    return prompt.map((message) => ({
+      ...message,
+      content: processContent(message.content),
+    }));
+  }
+};
+
+const validateDatasetItem = (
+  itemInput: Prisma.JsonValue,
+  variables: string[],
+): itemInput is Prisma.JsonObject => {
+  if (!isValidPrismaJsonObject(itemInput)) {
+    return false;
+  }
+  return variables.some(
+    (variable) =>
+      Object.keys(itemInput).includes(variable) &&
+      typeof itemInput[variable] === "string",
+  );
+};
+
+export const createExperimentJob = async ({
+  event,
+}: {
+  event: z.infer<typeof ExperimentCreateEventSchema>;
+}) => {
+  logger.info("Processing experiment create job", event);
+  const { datasetId, projectId, runId } = event;
+
+  /********************
+   * INPUT VALIDATION *
+   ********************/
+
+  // first, fetch all the context required for the experiment
+  const datasetRun = await kyselyPrisma.$kysely
+    .selectFrom("dataset_runs")
+    .selectAll()
+    .where("id", "=", runId)
+    .where("project_id", "=", projectId)
+    .executeTakeFirstOrThrow();
+
+  if (!datasetRun.metadata) {
+    throw new ForbiddenError(
+      "Langfuse in-app experiments can only be run with available model and prompt configurations.",
+    );
+  }
+
+  const metadata = ExperimentMetadataSchema.safeParse(datasetRun.metadata);
+  if (!metadata.success) {
+    throw new ForbiddenError(
+      "Langfuse in-app experiments can only be run with available model and prompt configurations.",
+    );
+  }
+
+  // validate the prompt
+  const { prompt_id, provider, model, model_params } = metadata.data;
+
+  const prompt = await kyselyPrisma.$kysely
+    .selectFrom("prompts")
+    .selectAll()
+    .where("id", "=", prompt_id)
+    .where("project_id", "=", event.projectId)
+    .executeTakeFirstOrThrow();
+
+  if (!prompt) {
+    logger.error(`Prompt ${prompt_id} not found for project ${projectId}`);
+    throw new InvalidRequestError(
+      `Prompt ${prompt_id} not found for project ${projectId}`,
+    );
+  }
+
+  const validatePromptContent = PromptContentSchema.safeParse(prompt.prompt);
+
+  if (validatePromptContent.error) {
+    logger.error(
+      `Prompt content not in expected format ${prompt_id} not found for project ${projectId}`,
+    );
+    throw new InternalServerError(
+      `Prompt ${prompt_id} not found in expected format for project ${projectId}`,
+    );
+  }
+
+  const extractedVariables = extractVariables(
+    prompt?.type === "text"
+      ? (prompt.prompt?.toString() ?? "")
+      : JSON.stringify(prompt.prompt),
+  );
```

Comment:
Here is another case where we attempt to parse the {{ }} notation. IMO a library would be useful here, or we create a global, internal util.

Reply:
That's fair. The function to `extractVariables` is used quite consistently throughout the code base, though it could do with tests and possibly the use of a library. Have added backlog ticket for myself to fix eventually 🙌